shared_stack.bash

```diff
@@ -0,0 +1,214 @@
+# Science Pipelines.
+#
+# End users looking to install the stack for personal use should prefer the
+# procedure described in pipelines.lsst.io. Developers looking to build the
+# latest versions of the LSST code should prefer the lsstsw build tool.
+#
+# Specifically, we:
+#
+# - Get the latest version of the lsstinstall tool into the STATE directory;
+# - Get the list of tags from the EUPS distribution server;
+# - For each tag which is not in the list of already-installed-and-deleted tags
+#   in "tags.deleted" and which does not have a directory under $ROOT/tag:
+#   - use lsstinstall to install the appropriate rubin-env conda environment if
+#     needed
+#   - add useful additional packages for developers if necessary
+#   - use "eups distrib install" to install the Science Pipelines from binary
+#     artifacts for the top-level PRODUCT
+#   - create a loadLSST.sh script for users to load the environment from
+#     scratch, a switch.sh script to activate the environment if conda is
+#     already loaded, and an env_name file to simplify other tooling that may
+#     need to map tags to rubin-env versions;
+# - Maintain symlinks for the latest daily and weekly tags;
+# - Maintain a common base loadLSST.sh script;
+# - For each tag beyond the number to be kept for dailies (KEEP_d) and
+#   weeklies (KEEP_w):
+#   - add the tag to the "tags.deleted" list
+#   - temporarily move the $ROOT/tag subdirectory aside
+#   - untag (if shared) or remove (if exclusive) the eups packages with the tag
+#   - remove the $ROOT/tag subdirectory
+#
+# A -n option may be used to do a "dry run" that only prints out the commands
+# that would be executed.
+#
+# This code is intended to work properly if invoked multiple times in parallel,
+# e.g. if it takes longer than the interval between cron jobs, although there
+# is no resulting speedup.  Hence PID-labeled temporary directories and atomic
+# moves are used frequently.
+
+###############
+# CONFIGURATION
+###############
+
+# Software root directory
+ROOT=/scratch/rubinsw
+# State directory
+STATE=~/shared-stack
+# Top-level product to install
+PRODUCT=lsst_distrib
+# Number of daily releases to keep
+# shellcheck disable=SC2034
+KEEP_d=24
+# Number of weekly releases to keep
+# shellcheck disable=SC2034
+KEEP_w=26
+
+#######
+# Setup
+#######
+
+while getopts n opt; do
+  case "$opt" in
+    n) dryrun="echo"
+       ;;
+    *) echo "Usage: $0 [-n]"
+       exit 1 ;;
+  esac
+done
+[ $OPTIND -le $# ] && echo "Usage: $0 [-n]" && exit 1
+# Ensure that there's no contamination from a pre-existing environment
+unset LSST_CONDA_ENV_NAME PYTHONPATH
+
+##############
+# Installation
+##############
+
+set -e
+mkdir -p "$STATE"
+cd "$STATE"
+# Label temporaries with our unique PID
+tmp=tmp.$$
+mkdir -p "$tmp" "$ROOT/tag"
+[ -e tags.deleted ] || touch tags.deleted
+umask 022
+
+# Get list of tags
+$dryrun curl -L https://ls.st/lsstinstall -o $tmp/lsstinstall
+$dryrun mv $tmp/lsstinstall lsstinstall
```

Comment:
That's OK.  The shell keeps the file descriptor open to the previous version, so it will continue to read it even though the file is unlinked.  The file goes away when all open descriptors are closed.